core/frontend/src/tile/map/EsriOAuth2.ts

```diff
@@ -0,0 +1,193 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { assert, BeEvent } from "@bentley/bentleyjs-core";
+import { ArcGisOAuth2Token, ArcGisTokenManager, ArcGisUtilities, MapLayerTokenEndpoint} from "../internal";
+
+/** @internal */
+export enum EsriOAuth2EndpointType {Authorize,Token}
+
+/** @internal */
+export class EsriOAuth2Endpoint implements MapLayerTokenEndpoint {
```

Comment:
I think it would be better to put all of the OAuth pieces into the map layers widget directly alongside the callback. Then the frontend code would just take the data it needs to communicate (i.e. access token plus any other info).

I'd have to dig a bit more for the appropriate interface but possibly one that would work for both Online and Enterprise workflows.

Reply:
I would argue that all imagery providers (except maybe Bing?) should be moved outside core, not only the Oauth2 stuff.  We are lacking an interface to achieve that.

Reply:
fair enough :) if the plan is to do that and all of this mostly stays internal we can leave it as is for now but maybe we should chat about the larger update at some point soon.

Reply:
I dont mind refactoring ImageryProviders prior to Oauth2... we simply need a way to inject providers when IModelApp.

Reply:
would they be needed in IModelApp initialize or TileAdmin initialize?

Reply:
`IModelApp.mapLayerFormatRegistry.register` can be called to register a maplayer format.